presto-tests/src/main/java/com/facebook/presto/tests/statistics/InMemoryHistoryBasedPlanStatisticsProvider.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tests.statistics;
+
+import com.facebook.presto.spi.plan.PlanNodeWithHash;
+import com.facebook.presto.spi.statistics.HistoricalPlanStatistics;
+import com.facebook.presto.spi.statistics.HistoryBasedPlanStatisticsProvider;
+import com.facebook.presto.spi.statistics.PlanStatistics;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static org.testng.Assert.assertTrue;
+
+public class InMemoryHistoryBasedPlanStatisticsProvider
+        implements HistoryBasedPlanStatisticsProvider
+{
+    private final Map<String, HistoricalPlanStatistics> cache = new ConcurrentHashMap<>();
+    private final Semaphore semaphore = new Semaphore(1);
```

Comment:
I know this is just a test class, but I'm confused about this locking setup.
1. Why use a semaphore with 1 permit instead of making the putStats() method synchronized?
2. Why do we need any kind of locks since the cache is a ConcurrentHashMap?
3. I'm confused about acquiring a lock in the constructor.  It seems weird to acquire a lock in the constructor, then something in the same thread needs to putStats in order to free it up, but then other threads might access it too?
4. Why use any kind of concurrency instead of creating a new provider for each need.  otherwise the cache will just keep growing as more tests are added.

Reply:
Since query events are processed in a separate thread, query finishing doesn't mean these events are processed. The semaphore is to make sure we wait till the event is processed and then go ahead with next query which we can use these stats

Reply:
Also for it to work, the tests accessing it need to be single threaded, which I have marked. Didn't put much thought into design as it was a test

Reply:
Makes sense.  Maybe add a comment.